### PR TITLE
Handle gameplay clock's "true gameplay rate" while GCC is stopped

### PR DESCRIPTION
**Pull request by frenzibyte**
- 1 out of 2 for resolving https://github.com/ppy/osu/issues/14162.

Previously when the underlying clocks of the GCC have a rate of 0 (by the ["pause frequency adjustment"](https://github.com/ppy/osu/blob/4b55ba7fa4d9a93ac58ef75941f2a41a17636dc2/osu.Game/Screens/Play/MasterGameplayClockContainer.cs#L99-L112)), the `TrueGameplayRate` property can't do anything but incorrectly return 0, while it's supposed not to, as the adjustment behind 0 is a "non-gameplay adjustment".

I've instead removed the whole `NonGameplayAdjustments` logic and flipped things around, by making `MasterGameplayClockContainer` itself accept adjustments and consider them as "gameplay adjustments", and pass the MGCC to mods as an `ITrack` with the adjustments delegated to the "gameplay adjustments" holder. (I was hoping to avoid doing that but it's apparently something mods like `ModTimeRamp` need for finding the track's raw current time)

Along with that, `HardwareCorrectionOffsetClock` now uses a newly introduced "`PlaybackRate`" property to scale the offset properly without being affected by the pause adjustment.

All of these changes have test coverage associated with them, which would also help explaining things more easily.

### CODE REVIEW

File: osu.Game/Screens/Play/MasterGameplayClockContainer.cs [hunk 1]
```diff
@@ -43,7 +42,23 @@ public class MasterGameplayClockContainer : GameplayClockContainer
...
+        /// <summary>
+        /// The rate of gameplay when playback is at 100%.
+        /// This excludes any seeking / user adjustments.
+        /// </summary>
+        public double TrueGameplayRate => GameplayTrack.AggregateFrequency.Value * GameplayTrack.AggregateTempo.Value;
+
+        /// <summary>
+        /// The true gameplay rate combined with the <see cref="UserPlaybackRate"/> value.
+        /// </summary>
+        public double PlaybackRate => TrueGameplayRate * UserPlaybackRate.Value;
```
Comment (bdach) [inline osu.Game/Screens/Play/MasterGameplayClockContainer.cs:59]: what does this *mean* in human words? this xmldoc merely restates the code.

**Comment by bdach** [comment]
i don't know if i'm just dense but i don't get any of this, or what it has to do with the original issue.

my intuition tells me that this diff _may_ indicate that offsets and playback rates are doubly applied to master and child clock while they should probably be only applied to the master clock? but i sincerely am unable to properly understand any of this from either the code or the description.

**Comment by frenzibyte** [comment]
The test cases should have been quite clear about this, but it's pretty much that a stopped master clock with audio offsets applied could mess up the synchronization because of the pause adjustment making Rate = 0.

I am differentiating all of these rates to:
 - Rate: the aggregate rate of basically everything, and it is how components should animate visually
- TrueGameplayRate: the rate of gameplay itself away from outside playback adjustments, used for spm calculation
 - PlaybackRate: the current exact rate of gameplay playback, away from pause frequency adjustment, used for hardware offset clock to scale up properly, without being affected by pausing.

Naming open for suggestions, I'll rewrite the xmldocs.

File: osu.Game/Screens/Play/MasterGameplayClockContainer.cs (hunk 1 above)
Comment (frenzibyte) [reply osu.Game/Screens/Play/MasterGameplayClockContainer.cs]: Yeah my bad, I was planning to rewrite the xmldocs but forgot that while I was resolving the side issues pointed by the tests, I've explained enough now, feel free to reword or point out anything vague with it.

**Comment by peppy** [comment]
This PR can't exist in its current fashion. It's piling more complexity on an already complex system. It looks like you're trying to achieve something similar to what I fixed in https://github.com/ppy/osu/pull/14776 - maybe my PR will give you some ideas on how we can better handle this without adding the extra nested/redirection complexity you have here?

File: osu.Game.Tests/Gameplay/TestSceneMasterGameplayClockContainer.cs [hunk 2]
```diff
@@ -54,5 +60,134 @@ public void TestElapseThenReset()
...
+
+                gcc.Stop();
+                gcc.Reset();
+            });
+
+            AddStep("set audio offset to 100ms", () => config.SetValue(OsuSetting.AudioOffset, 100.0));
+
+            AddAssert("current time has 100% offset", () => gcc.GameplayClock.CurrentTime == 100);
+
+            AddStep("add 0 track freq adjustment", () => working.Track.AddAdjustment(AdjustableProperty.Frequency, new BindableDouble()));
```
Comment (peppy) [inline osu.Game.Tests/Gameplay/TestSceneMasterGameplayClockContainer.cs:123]: I was hoping to use your tests against my PR to see if I can get things working, but this seems wrong? A mod could potentially be adjusting the frequency of the track in a similar way, and *should* be considered as a rate adjust in that case?

Comment (frenzibyte) [reply osu.Game.Tests/Gameplay/TestSceneMasterGameplayClockContainer.cs:123]: The behaviours in general have changed in this PR, and the *applicable* adjustments should go to `gcc.GameplayTrack`. This case in this PR is identical to how `pauseFreqAdjust` gets applied.

**Comment by frenzibyte** [comment]
> This PR can't exist in its current fashion. It's piling more complexity on an already complex system. It looks like you're trying to achieve something similar to what I fixed in #14776 - maybe my PR will give you some ideas on how we can better handle this without adding the extra nested/redirection complexity you have here?

The general idea of this PR is to separate non-gameplay adjustments from gameplay adjustments properly by having a separate `AudioAdjustment` for the gameplay, and wrapping it in an `ITrack` considered as the "gameplay track".

While it may be a bit of a complicated solution, I hoped to solve it in a way that doesn't involve relying on certain adjustments *arbitrarily* (and get rid of `NonGameplayAdjustments` with MGCC's "gameplay adjustments" instead, which `TrueGameplayRate` would take the rate from instead of doing the inverse with `NonGameplayAdjustments`).

I'm interested in hearing what the others think about this as well, and whether #14776 would remain to be a final solution for this, as I don't really have any other idea towards this at the moment.

**Pull request closed without merging**